src/main/java/com/google/api/gax/grpc/OperationFuture.java

```diff
@@ -0,0 +1,372 @@
+        responseClass,
+        pollingInterval,
+        new Waiter());
+  }
+
+  // package-private for testing
+  @VisibleForTesting
+  static <ResponseT extends Message> OperationFuture<ResponseT> create(
+      OperationsApi operationsApi,
+      ListenableFuture<Operation> startOperationFuture,
+      ScheduledExecutorService executor,
+      Class<ResponseT> responseClass,
+      Duration pollingInterval,
+      Waiter waiter) {
+    SettableFuture<ResponseT> finalResultFuture = SettableFuture.create();
+    CountDownLatch asyncCompletionLatch = new CountDownLatch(1);
+    Future<ResponseT> dataGetterFuture =
+        executor.submit(
+            new DataGetterRunnable<ResponseT>(
+                startOperationFuture,
+                finalResultFuture,
+                operationsApi,
+                responseClass,
+                pollingInterval,
+                waiter,
+                asyncCompletionLatch));
+    OperationFuture<ResponseT> operationFuture =
+        new OperationFuture<>(
+            startOperationFuture, finalResultFuture, dataGetterFuture, asyncCompletionLatch);
+    return operationFuture;
+  }
+
+  private OperationFuture(
+      ListenableFuture<Operation> startOperationFuture,
+      SettableFuture<ResponseT> finalResultFuture,
+      Future<ResponseT> dataGetterFuture,
+      CountDownLatch asyncCompletionLatch) {
+    this.startOperationFuture = startOperationFuture;
+    this.finalResultFuture = finalResultFuture;
+    this.dataGetterFuture = dataGetterFuture;
+    this.asyncCompletionLatch = asyncCompletionLatch;
+  }
+
+  private static class DataGetterRunnable<ResponseT extends Message>
+      implements Callable<ResponseT> {
+    private final ListenableFuture<Operation> startOperationFuture;
+    private final SettableFuture<ResponseT> finalResultFuture;
+    private final OperationsApi operationsApi;
+    private final Class<ResponseT> responseClass;
+    private final Duration pollingInterval;
+    private final Waiter waiter;
+    private final CountDownLatch asyncCompletionLatch;
+
+    public DataGetterRunnable(
+        ListenableFuture<Operation> startOperationFuture,
+        SettableFuture<ResponseT> finalResultFuture,
+        OperationsApi operationsApi,
+        Class<ResponseT> responseClass,
+        Duration pollingInterval,
+        Waiter waiter,
+        CountDownLatch asyncCompletionLatch) {
+      this.startOperationFuture = startOperationFuture;
+      this.finalResultFuture = finalResultFuture;
+      this.operationsApi = operationsApi;
+      this.responseClass = responseClass;
+      this.pollingInterval = pollingInterval;
+      this.waiter = waiter;
+      this.asyncCompletionLatch = asyncCompletionLatch;
+    }
+
+    @Override
+    public ResponseT call() {
+      try {
+        callImpl();
+      } finally {
+        asyncCompletionLatch.countDown();
+      }
+      return null;
+    }
+
+    public void callImpl() {
+      Operation firstOperation = null;
```

Comment:
Do we need to keep separate firstOperation and latestOperation? It seems like a strange behavior to change the name of the operation midway.